app/questionnaire/questionnaire_schema.py

```diff
@@ -215,19 +215,36 @@ def get_post_submission(self) -> ImmutableDict:
         schema: ImmutableDict = self.json.get("post_submission", ImmutableDict({}))
         return schema
 
+    def _check_list_exists(self, rules: Union[dict, Sequence], list_name: str) -> bool:
+        # old rules
+        if isinstance(rules, Sequence):
+            return any(
+                rule.get("list") == list_name and "list" in rule for rule in rules
+            )
+
+        # New rules
+        if isinstance(rules, dict) and rules.get("source") == "list":
+            return rules.get("identifier") == list_name
```

Comment:
This would actually still fail, as if source is not list, we still don't need to iterate any further. Should be something like below but may need rework based on the comment below.

```py
        if isinstance(rules, dict) and (source := rules.get("source")):
            return source == "list" and rules.get("identifier") == list_name
```